### DISCUSSION TITLE
How to use "file_socket" parameter in WebServerControl? (Any example inputs)

### DISCUSSION THREAD

**Discussion by yifanm4** [idaholab/moose#27975] [Q&A Modules: General] [answered]
Hi MOOSE developers, 

I want to use WebServerControl to make MOOSE communicate with my Open Modelica application for research purpose. I saw in the document that it's required to use either "port" parameter or "file_socket" parameter so the server can listen to the port. I looked into the example input files and none of them actually implemented any of them. I was wondering if anybody can teach me the general work flow of this? 

I was trying to follow the steps listed in  https://mooseframework.inl.gov/python/MooseControl/index.html to get a taste of it. But I cannot seem to get it working since I don't know how to set my file_socket parameter. 

Thanks! 
Yifan

**Comment by GiudGiud** [comment]
Hello

This is a very recent feature. Which version of moose (which commit) are you using?

Guillaume

**Reply by yifanm4** [reply]
Hello, 

I am using the newest version (2024.06.17)

**Reply by GiudGiud** [reply]
@loganharbour

**Reply by loganharbour** [reply, marked as answer]
If you're using the MooseControl python object, you don't need to set the port or file socket. It's done for you internally by the script. 

Within the MooseControl object, you specify the name of the WebServerControl object in your input file. It uses this name to take care of the port/file socket for you via a command line argument.

**Reply by yifanm4** [reply]
Thank you for the quick response! It does make more sense now. But I come into an error message that I don't know how to fix, can you please help out or point to the resources that I can look at? 

**I create a python script similar to what is shown in the example:** 



from MooseControl import MooseControl

command = 'mpiexec -n 2 ./ex01_inputfile/ex01-opt -i /ex01_inputfile/ex01_withWebControl.i'
command_list = command.split()
control = MooseControl(moose_command=command_list,
                       moose_control_name='web_server')
control.initialize()



But I got the following error messages when I run it:

```MooseControl: Determined file socket /var/folders/b8/pz63xbjs2dd5df815t7c68fc0000gn/T/moose_control_i1aupz for communication
MooseControl: Spawning MOOSE with command "['mpiexec', '-n', '2', './ex01_inputfile/ex01-opt', '-i', '/ex01_inputfile/ex01_withWebControl.i', 'Controls/web_server/file_socket=/var/folders/b8/pz63xbjs2dd5df815t7c68fc0000gn/T/moose_control_i1aupz']"
MooseControl: Waiting for the webserver to start on "http+unix:///var/folders/b8/pz63xbjs2dd5df815t7c68fc0000gn/T/moose_control_i1aupz"
Traceback (most recent call last):
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/encodings/idna.py", line 165, in encode
    raise UnicodeError("label too long")
UnicodeError: label too long

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/yifanmao/projects/moose/examples/ex01_inputfile/ex01_webControl.py", line 11, in <module>
    control.initialize()
  File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 286, in initialize
    if self.isListening():
  File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 213, in isListening
    r = self._requests_get(f'{self._url}/check')
  File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 142, in _requests_get
    return self._requests_wrapper('get', *args, **kwargs)
  File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 138, in _requests_wrapper
    return getattr(accessor, function_name)(*args, **kwargs)
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/sessions.py", line 602, in get
    return self.request("GET", url, **kwargs)
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/sessions.py", line 579, in request
    settings = self.merge_environment_settings(
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/sessions.py", line 760, in merge_environment_settings
    env_proxies = get_environ_proxies(url, no_proxy=no_proxy)
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/utils.py", line 832, in get_environ_proxies
    if should_bypass_proxies(url, no_proxy=no_proxy):
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/utils.py", line 816, in should_bypass_proxies
    bypass = proxy_bypass(parsed.hostname)
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/urllib/request.py", line 2654, in proxy_bypass
    return proxy_bypass_macosx_sysconf(host)
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/urllib/request.py", line 2631, in proxy_bypass_macosx_sysconf
    return _proxy_bypass_macosx_sysconf(host, proxy_settings)
  File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/urllib/request.py", line 2599, in _proxy_bypass_macosx_sysconf
    hostIP = socket.gethostbyname(hostonly)
UnicodeError: encoding with 'idna' codec failed (UnicodeError: label too long)```

Can you please give me some suggestions on how to fix it?

**Reply by yifanm4** [reply]
> Thank you for the quick response! It does make more sense now. But I come into an error message that I don't know how to fix, can you please help out or point to the resources that I can look at?
> 
> **I create a python script similar to what is shown in the example:**
> 
> from MooseControl import MooseControl
> 
> command = 'mpiexec -n 2 ./ex01_inputfile/ex01-opt -i ./ex01_inputfile/ex01_withWebControl.i' command_list = command.split() control = MooseControl(moose_command=command_list, moose_control_name='web_server') control.initialize()
> 
> But I got the following error messages when I run it:
> 
> ```
> MooseControl: Spawning MOOSE with command "['mpiexec', '-n', '2', './ex01_inputfile/ex01-opt', '-i', '/ex01_inputfile/ex01_withWebControl.i', 'Controls/web_server/file_socket=/var/folders/b8/pz63xbjs2dd5df815t7c68fc0000gn/T/moose_control_i1aupz']"
> MooseControl: Waiting for the webserver to start on "http+unix:///var/folders/b8/pz63xbjs2dd5df815t7c68fc0000gn/T/moose_control_i1aupz"
> Traceback (most recent call last):
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/encodings/idna.py", line 165, in encode
>     raise UnicodeError("label too long")
> UnicodeError: label too long
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/Users/yifanmao/projects/moose/examples/ex01_inputfile/ex01_webControl.py", line 11, in <module>
>     control.initialize()
>   File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 286, in initialize
>     if self.isListening():
>   File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 213, in isListening
>     r = self._requests_get(f'{self._url}/check')
>   File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 142, in _requests_get
>     return self._requests_wrapper('get', *args, **kwargs)
>   File "/Users/yifanmao/projects/moose/python/MooseControl/MooseControl.py", line 138, in _requests_wrapper
>     return getattr(accessor, function_name)(*args, **kwargs)
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/sessions.py", line 602, in get
>     return self.request("GET", url, **kwargs)
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/sessions.py", line 579, in request
>     settings = self.merge_environment_settings(
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/sessions.py", line 760, in merge_environment_settings
>     env_proxies = get_environ_proxies(url, no_proxy=no_proxy)
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/utils.py", line 832, in get_environ_proxies
>     if should_bypass_proxies(url, no_proxy=no_proxy):
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/site-packages/requests/utils.py", line 816, in should_bypass_proxies
>     bypass = proxy_bypass(parsed.hostname)
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/urllib/request.py", line 2654, in proxy_bypass
>     return proxy_bypass_macosx_sysconf(host)
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/urllib/request.py", line 2631, in proxy_bypass_macosx_sysconf
>     return _proxy_bypass_macosx_sysconf(host, proxy_settings)
>   File "/Users/yifanmao/mambaforge3/envs/moose/lib/python3.10/urllib/request.py", line 2599, in _proxy_bypass_macosx_sysconf
>     hostIP = socket.gethostbyname(hostonly)
> UnicodeError: encoding with 'idna' codec failed (UnicodeError: label too long)```
> 
> Can you please give me some suggestions on how to fix it? 
> ```

I actually found a way to go around It. I can set my temp file directory to a shorter path (https://stackoverflow.com/questions/847850/cross-platform-way-of-getting-temp-directory-in-python)
the default temp dir on macOS that is grabbed by tempfile.gettempdir() seems to be too long

I tried ```mkdir /Users/yifanmao/tmp``` and then ```export TMPDIR=/Users/yifanmao/tmp``` before I run the script and it works as expected. 

Please let me know if there are better ways to handle it. Thanks

**Reply by GiudGiud** [reply]
I'm glad you figured it out.
The error message is clear enough imo, I dont think we need a corrective action for this, but @loganharbour will decide

**Reply by loganharbour** [reply]
Yeah, unfortunately this is an issue I've come across on macs (not liking long directory paths).

I'm not sure if there's a good way around it unfortunately. I could make it the same directory that your input file is in, but that path might be long too!
